File: dotnet/runtime/Metamodel/Representations/P6list.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Rakudo.Metamodel.Representations
{
    /// <summary>
    /// This is a very first cut at a list representation. Essentially,
    /// it just knows how to store a list of objects at the moment. At
    /// some point we need to define the way that it will handle compact
    /// arrays.
    /// </summary>
    public class P6list : Representation
    {
        internal class Instance : RakudoObject
        {
            /// <summary>
            /// Just use a .Net List at the moment, but an array would
            /// be more efficient in the long run (though give us more
            /// stuff to implement ourselves).
            /// </summary>
            public List<RakudoObject> Storage;
            public Instance(SharedTable STable)
            {
                this.STable = STable;
            }
        }

        /// <summary>
        /// Create a new type object.
        /// </summary>
        /// <param name="HOW"></param>
        /// <returns></returns>
        public override RakudoObject type_object_for(RakudoObject MetaPackage)
        {
            var STable = new SharedTable();
            STable.HOW = MetaPackage;
            STable.REPR = this;
            STable.WHAT = new Instance(STable);
            return STable.WHAT;
        }

        /// <summary>
        /// Creates an instance of the type with the given type object.
        /// </summary>
        /// <param name="WHAT"></param>
        /// <returns></returns>
        public override RakudoObject instance_of(RakudoObject WHAT)
        {
            var Object = new Instance(WHAT.STable);
            Object.Storage = new List<RakudoObject>();
            return Object;
        }

        /// <summary>
        /// Determines if the representation is defined or not.
        /// </summary>
        /// <param name="Obj"></param>
        /// <returns></returns>
        public override bool defined(RakudoObject Obj)
        {
            return ((Instance)Obj).Storage != null;
        }

        public override RakudoObject get_attribute(RakudoObject Object, RakudoObject ClassHandle, string Name)
        {
            throw new NotImplementedException();
        }

        public override RakudoObject get_attribute_with_hint(RakudoObject Object, RakudoObject ClassHandle, string Name, int Hint)
        {
            throw new NotImplementedException();
        }

        public override void bind_attribute(RakudoObject Object, RakudoObject ClassHandle, string Name, RakudoObject Value)
        {
            throw new NotImplementedException();
        }

        public override void bind_attribute_with_hint(RakudoObject Object, RakudoObject ClassHandle, string Name, int Hint, RakudoObject Value)
        {
            throw new NotImplementedException();
        }

        public override int hint_for(RakudoObject ClassHandle, string Name)
        {
            throw new NotImplementedException();
        }

        public override void set_int(RakudoObject Object, int Value)
        {
            throw new NotImplementedException();
        }

        public override int get_int(RakudoObject Object)
        {
            throw new NotImplementedException();
        }

        public override void set_num(RakudoObject Object, double Value)
        {
            throw new NotImplementedException();
        }

        public override double get_num(RakudoObject Object)
        {
            throw new NotImplementedException();
        }

        public override void set_str(RakudoObject Object, string Value)
        {
            throw new NotImplementedException();
        }

        public override string get_str(RakudoObject Object)
        {
            throw new NotImplementedException();
        }
    }
}

File: java/runtime/Rakudo/Metamodel/Representations/P6list.java
package Rakudo.Metamodel.Representations;

import java.util.ArrayList;
//import Rakudo.Metamodel.Hints;
import Rakudo.Metamodel.RakudoObject;
import Rakudo.Metamodel.Representation;
import Rakudo.Metamodel.SharedTable;
import Rakudo.Runtime.ThreadContext;
import Rakudo.Serialization.SerializationContext;

/// <summary>
/// This is a very first cut at a list representation. Essentially,
/// it just knows how to store a list of objects at the moment. At
/// some point we need to define the way that it will handle compact
/// arrays.
/// </summary>
public class P6list implements Representation
{
    public class Instance extends RakudoObject
    {
        /// <summary>
        /// Just use a .Net List at the moment, but an array would
        /// be more efficient in the long run (though give us more
        /// stuff to implement ourselves).
        /// </summary>
        public ArrayList<RakudoObject> Storage;
        public Instance(SharedTable sharedTable)
        {
            this.setSTable(sharedTable);
        }
    }

    /// <summary>
    /// Create a new type object.
    /// </summary>
    /// <param name="HOW"></param>
    /// <returns></returns>
    public RakudoObject type_object_for(ThreadContext tc, RakudoObject metaPackage)
    {
        SharedTable sharedTable = new SharedTable();
        sharedTable.HOW = metaPackage;
        sharedTable.REPR = this;
        sharedTable.WHAT = new Instance(sharedTable);
        return sharedTable.WHAT;
    }

    /// <summary>
    /// Creates an instance of the type with the given type object.
    /// </summary>
    /// <param name="WHAT"></param>
    /// <returns></returns>
    public  RakudoObject instance_of(ThreadContext tc, RakudoObject WHAT)
    {
        Instance object = new Instance(WHAT.getSTable());
        object.Storage = new ArrayList<RakudoObject>();
        return object;
    }

    /// <summary>
    /// Determines if the representation is defined or not.
    /// </summary>
    /// <param name="Obj"></param>
    /// <returns></returns>
    public boolean defined(ThreadContext tc, RakudoObject object)
    {
        return ((Instance)object).Storage != null;
    }

    public  RakudoObject get_attribute(ThreadContext TC, RakudoObject object, RakudoObject classHandle, String name)
    {
        throw new UnsupportedOperationException();
    }

    public  RakudoObject get_attribute_with_hint(ThreadContext TC, RakudoObject object, RakudoObject classHandle, String name, int hint)
    {
        throw new UnsupportedOperationException();
    }

    public  void bind_attribute(ThreadContext TC, RakudoObject object, RakudoObject classHandle, String name, RakudoObject value)
    {
        throw new UnsupportedOperationException();
    }

    public  void bind_attribute_with_hint(ThreadContext tc, RakudoObject object, RakudoObject classHandle, String name, int hint, RakudoObject value)
    {
        throw new UnsupportedOperationException();
    }

    public  int hint_for(ThreadContext tc, RakudoObject classHandle, String name)
    {
        throw new UnsupportedOperationException();
    }

    public  void set_int(ThreadContext tc, RakudoObject object, int value)
    {
        throw new UnsupportedOperationException();
    }

    public  int get_int(ThreadContext tc, RakudoObject object)
    {
        throw new UnsupportedOperationException();
    }

    public  void set_num(ThreadContext tc, RakudoObject object, double value)
    {
        throw new UnsupportedOperationException();
    }

    public  double get_num(ThreadContext tc, RakudoObject object)
    {
        throw new UnsupportedOperationException();
    }

    public  void set_str(ThreadContext tc, RakudoObject object, String value)
    {
        throw new UnsupportedOperationException();
    }

    public  String get_str(ThreadContext tc, RakudoObject object)
    {
        throw new UnsupportedOperationException();
    }
}

